### PR TITLE
[CR] Salvaging mechanic rework proposal

### PR DESCRIPTION
**Pull request by LcsTen**
#### Summary

SUMMARY: Balance "Modify salvaging algorithm to be mass-based"

#### Purpose of change

Change the salvaging algorithm to be mass-based rather than volume-based. #37063.

#### Describe the solution

The two ideas are:
* The salvaging algorithm must be mass-based rather than volume-based (obviously)
* We should avoid using arbitrary values such as the "minimal_volume_to_cut" and use the resulting components weight instead.

The new algorithm looks like:
(I'm taking the rubber dog rainsuit as an example)
1. Get the list of the materials of the item (neoprene, plastic, rubber)
2. Evenly assign the weight of the item to each material (1960g/3 = 653g)
3. Compute the maximum salvaged components for each material by dividing by the mass of one unit of that component
   * Neoprene -> neoprene patches (80g) -> 653/80 = 8
   * Plastic -> plastic chunks (50g) -> 653/50 = 13
   * Rubber -> chunks of rubber (38g) -> 653/38 = 17
4. Compute how many components will be lost (the algorithm is the same as before: entropy, dexterity roll and damaged loss)
5. For each part lost, randomly choose which component is lost

#### Describe alternatives you've considered

I'm open to any improvements for the new algorithm.

I think myself the components loss shouldn't be count-based since it means heavy items are less affected while light items may easily lose 100% of their components.
Instead, the loss should be relative to the weight of the item, for example, each components loss roll may delete some percentage of the effectively salvageable mass of the item, by multiplying the weight by 0.9 for example, and only then we may compute the possible components on the final value of the effectively salveagable mass.

Ideally the weight of the item shouldn't be distributed evenly, but then we would have to describe how much of each material is present in the item, by replacing ``[ "neoprene", "plastic", "rubber" ]`` by  something like ``[ [ 0.3, "neoprene" ], [ 0.5, "plastic" ], [ 0.2, "rubber" ] ]``. It's probably out of scope though.

#### Testing

An inexperienced character will usually lose from 0 to 3 components when trying to salvage an undamaged item.

Try to salvage the rubber dog rainsuit, the biosilicified chitin dog mesh harness, the chitin dog mesh harness, the kevlar dog harness and the leather dog harness and make sure the resulting components are at most the value shown in the examples of the Describe the solution and Additional context sections.
Try to salvage the laundry basket and make sure it yields at most **13** plastic chunks (``680/50``).

#### Additional context

I tested this new salvage algorithm on dog armors, and the result aren't really satisfying.
Some examples:

* Biosilicified chitin dog mesh harness costs **12** chunks of biosilicified chitin to craft, the disassembly recipe yields **6** chunks. The new salvage algorithm yields **48** chunks (``5362/111``).
* Chitin dog mesh harness costs **12** chunks of chunks of chitin to craft, the disassembly recipe yields **6** chunks. The new salvage algorithm yields **60** chunks (``5362/89``).
* The disassembly recipe of the Kevlar dog harness yields **6** layered Kevlar panels. The new salvage algorithm yields **53** panels (``4290/80``).
* The disassembly recipe of the leather dog harness yields **4** leather patches. The new salvage algorithm yields **21** patches (``2145/100``).

Those results imply one or more of the following:
* This new algorithm is wrong.
* The components are too light.
* The items are too heavy.
* The crafting recipes use too few components.
* We should rely on a constant, arbitrary value rather than on components weight

Edit: Updated the Testing section.

### CODE REVIEW

**Comment by anothersimulacrum** [comment]
I think this is definitely an improvement over the volume based system, though we'll probably be doing some cleaning up for a while. 
Related: #42375,

**Comment by LcsTen** [comment]
Indeed, and I think part of the cleaning must come with this PR. The problem is, we will have to dose carefully if when we have to make components heavier, make items lighter and increase components used by recipes.
I'm currently working on a new option ``--check-recipes`` which check for the consistency of recipes weight-wise, it may make the task easier.

**Comment by anothersimulacrum** [comment]
A new command line option? Could I encourage you to integrate that into the JSON loading code, if it's not too difficult?

**Comment by LcsTen** [comment]
If my check is correct, there are 417 inconsistent recipes which must be cleaned.

**Comment by LcsTen** [comment]
I'm about to begin the cleaning. The rules I'm going to follow are:
1. If the recipe only has one component and no byproducts, both the result and the component will have ``(result.weight+component.weight)/2`` as their weight.
2. If the recipe has more than one component and no byproducts...
   1. If ``result.weight/components.weight`` < 1.1, result.weight = ``result.weight*(2-(result.weight/components.weight))``
   2. If ``result.weight/components.weight`` > 3, then the weight of the components and the number of components are multiplied by ``1+((result.weight/components.weight)-1)/3`` and the weight of the result is multiplied by ``1-((result.weight/components.weight)-1)/3``.
   3. Else, result.weight is multiplied by ``1-((result.weight/components.weight)-1)/2`` and the number of components are multiplied by ``1+((result.weight/components.weight)-1)/2``.
3. If the recipe has byproducts, remove enough byproducts to satisfy the condition. If the recipe has no more byproducts and still doesn't satisfy the condition, follow 2.

**Comment by jbytheway** [comment]
I propose the following change to help balance the unsatisfactory examples you quoted:

If the item can be disassembled, then recursively calculate the materials in each component, add those up and then use that as the distribution across materials, rather than assuming a uniform distribution.  This ought to give better results than a uniform distribution without having to go through every item and manually assign material ratios.

**Comment by LcsTen** [comment]
I see your point, but there are a two problems.

Firstly, I don't think that a item that can be salvaged can be disassembled once this PR is merged. The solution proposed by this PR is to replace the disassembly recipes by a more balanced salvaged system. Let's take the chitin dog mesh harness for example. It is currently not salvageable and has a disassembly recipe that yields 6 chunks, but it is an artificial way to fix an exploit that existed with the old salvage system. The idea is to remove that recipe and to make this item salvageable, without introducing an exploit.

Secondly, it won't solve the problem of the examples. I'm taking the chitin dog mesh harness as an example: the disassembly recipe yields 6 chunks of chitin and nothing else, which means the material distribution is 100% chitin. Indeed, as stated in the item description. Yet, the result still isn't satisfying.

I guess you are misunderstanding the problem I pointed out with my four unsatisfactory examples. The problem isn't the distribution of each material, since anyway these items have only one material (both as stated in the item description and in the disassembly recipe). The real problem here is that the components are too light and the results are too heavy.
Then again, I'm taking the chitin dog mesh harness as an example. It weighs 5362 g. To craft it, we need 12 chunks of chitin, each weighs 89 g.
That means the result weighs 5362 g while the components weigh 12*89 = 1068 g. In other words, the recipe creates 4294 g out of nowhere. This is the problem. Because now, my new system, which relies heavily on the mass of the item, will create 4294/89 = 48 chitin chunks out of nowhere. This is the cleaning anothersimulacrum spoke about: We need to enforce this weight consistency among the recipes of the game to avoid future salvage exploits.

**Comment by jbytheway** [comment]
Understood.  I see that I'm not solving the problem you are discussing.

I imagine there might still exist other items that exhibit the problem of material distribution, but I guess that's a problem for another PR.

**Comment by John-Candlebury** [comment]
I dont particularily mind that Aftershock recipes mismatch mass between total components and products.  The mod's generic-component crafting system is not intended to represent reality to any  degree of accuracy, and instead aims to balance crafting recipes based on the results utility and on how easy its to acquire the components. Most of your changes here would make the recipes much more harder to perform, in a way that disrupts the intended balance of the mod.

I'd like to ask you to revert the changes made to the Aftershock mod, and if you are adding a mass check to the games test suite, to blacklist the Aftershock directory from those tests.

**Comment by LcsTen** [comment]
I totally understand, I will revert the changes made on your mod.
The mass check isn't done during the test suite, but on recipe load, and as such it isn't possible to tell to which mod the recipe belongs.
Would you like the addition of a new flag such as "BYPASS_MASS_CHECK" to your mod recipes (among others), which would prevent the check from complaining?

**Comment by John-Candlebury** [comment]
Thank you!
Yes, a flag would be ok.

File: data/mods/Magiclysm/recipes/dragon_black.json [hunk 1]
```diff
@@ -2,7 +2,6 @@
   {
     "type": "recipe",
     "result": "black_dragon_tanning_hide",
-    "byproducts": [ [ "dragon_black_scale", 20 ] ],
```
Comment (KorGgenT) [inline data/mods/Magiclysm/recipes/dragon_black.json:5]: please don't remove this, i want you to be able to get scales from the hides.

File: data/mods/Magiclysm/recipes/recipe_potions.json [hunk 2]
```diff
@@ -36,7 +36,7 @@
     "result": "mana_potion_lesser",
     "qualities": [ { "id": "CONTAIN", "level": 1 }, { "id": "MANA_INFUSE", "level": 1 } ],
     "tools": [ [ [ "surface_heat", 10, "LIST" ] ], [ [ "rune_alchemist", -1 ] ] ],
-    "components": [ [ [ "lotus", 1 ] ], [ [ "potion_starter", 1 ] ] ],
```
Comment (KorGgenT) [inline data/mods/Magiclysm/recipes/recipe_potions.json:39]: all of these are magic potions so i'd appreciate if they could stay the same.

File: data/mods/Magiclysm/recipes/dragon_black.json (hunk 1 above)
Comment (LcsTen) [reply data/mods/Magiclysm/recipes/dragon_black.json:5]: I understand, should I keep the number or is it okay to remove some?

File: data/mods/Magiclysm/recipes/recipe_potions.json (hunk 2 above)
Comment (LcsTen) [reply data/mods/Magiclysm/recipes/recipe_potions.json:39]: I see, I will add the BYPASS_MASS_CHECK flag to the magic potions.

**Comment by I-am-Erk** [comment]
You seem to have your fingers in a lot of pies here. Why are you changing recipe calories and food masses in a salvage PR?

**Comment by LcsTen** [comment]
It is because I added a mass check on recipe load, which complains if the weight of the results is greater than the weight of the components to avoid salvage exploits.
Indeed, food shouldn't be salvaged and such a check isn't useful. I could revert the changes made on the food, and change the check to exclude either:
- recipes which use the cooking skill
- recipes whose result is food
- recipes in the CC_FOOD category

**Comment by LcsTen** [comment]
I removed the [WIP] tag as I consider the cleaning to be done.
For most recipes, this implies an increase in component usage. As such, the components become rarer. Maybe we should adapt the frequency of the components accordingly.
For some recipes, the change is huge. Maybe it would be preferable to revert those changes, and to plan it for another PR.

File: data/json/items/armor/torso_clothes.json [hunk 3]
```diff
@@ -726,7 +726,7 @@
     "type": "ARMOR",
     "name": { "str": "leather vest" },
     "description": "A vest made from thick leather.  Offers excellent protection from cuts.",
-    "weight": "997 g",
+    "weight": "1530670 mg",
```
Comment (I-am-Erk) [inline data/json/items/armor/torso_clothes.json:729]: You have way too many things here where you're changing from a reasonable gram value (or in this case one that should have just bean 1kg) and changing it to ridiculous milligram values

Comment (LcsTen) [reply data/json/items/armor/torso_clothes.json:729]: This value is the result of the cleaning of those two recipes:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/recipes/armor/torso.json#L1159
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/json/recipes/armor/pets_dog.json#L41

I understand that the precise miligram value is less readable than gram.
Would you like this value to be changed to 1530 g?

Comment (anothersimulacrum) [reply data/json/items/armor/torso_clothes.json:729]: We don't need milligram precision, but we should either adjust the recipe or just bump this is 1530g for weight accuracy.

Comment (I-am-Erk) [reply data/json/items/armor/torso_clothes.json]: Yes, I would say anything that is more than 5 grams should be rounded to the nearest gram, if you want to do it automatically.

**Pull request closed without merging**